### PR TITLE
`Adaptive learning`: Improve usability of competency selector 

### PR DESCRIPTION
**Pull request by milljoniaer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
This PR improves the UI of the competency selector within exercises & lecture units.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with some added competencies

1. Log in to Artemis
2. Edit an exercise and verify that it is possible to add + save a competency
3. Verify that it is possible to remove + save a competency

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [competency-selection.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6103/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/competency-selection/competency-selection.component.ts.html) | 97% | ✅ |

### Screenshots
image
image

## Summary by CodeRabbit

- **New Features**
    - Added `(ngModelChange)="cacheValidation()"` event binding to update cache on model change in quiz exercises.
    - Introduced a new competency selector component for more intuitive competency selection with checkboxes.
- **Style**
    - Removed the `class` attribute `mb-0` from a div element in the form group.
    - Added styling attributes like borders, border radius, and improved label display to the competency selector component.
- **Refactor**
    - Reorganized competency selection structure within the component for better efficiency.
- **Tests**
    - Updated tests for competency selector component, including competency selection, unselection, and onchange registration.

### CODE REVIEW

File: src/main/webapp/app/shared/competency-selection/competency-selection.component.scss [hunk 1]
```diff
@@ -0,0 +1,14 @@
+.competency-selector {
+    display: flex;
+    flex-direction: column;
+    flex-wrap: wrap;
+    column-gap: 24px;
+    max-height: 200px;
+    overflow-x: auto;
```
[the remark this answers is not in the document]
Comment (milljoniaer) [reply src/main/webapp/app/shared/competency-selection/competency-selection.component.scss:7]: Nope `x` is correct

File: src/main/webapp/app/shared/competency-selection/competency-selection.component.ts [hunk 2]
```diff
@@ -21,10 +22,11 @@ import { finalize } from 'rxjs';
...
     @Input() labelName: string;
     @Input() labelTooltip: string;
-    @Input() value: any;
+    // selected competencies
+    @Input() value: Competency[] = [];
     @Input() disabled: boolean;
-    @Input() error: boolean;
-    @Input() competencies: Competency[];
+    // all course competencies
+    @Input() competencies: Competency[] = [];
```
Comment (pzdr7) [inline src/main/webapp/app/shared/competency-selection/competency-selection.component.ts:29]: Tested on TS2 - I could not get any of the competencies to show up when editing the exercises. I suspect it has something to do with this line. 

The condition for loading the course competencies in `ngOnInit()` reads:

```ts
if(!this.competencies && courseId) {
  /* ... */
}
```

`!this.competencies` evaluates to `false` if `this.competencies` is an empty array, which seems to be the case when opening the edit view of an exercise.

**Review by pzdr7 — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/shared/competency-selection/competency-selection.component.ts (hunk 2 above)
Comment (milljoniaer) [reply src/main/webapp/app/shared/competency-selection/competency-selection.component.ts]: Yeah, you are right, I have overlooked this. 
It is fixed, please check again

**Review by pzdr7 — COMMENTED** [review]
Tested on TS2.

- Sign in as instructor > Course management > (Test course) > Exercises > Edit an exercise > I can add and remove competencies, and the changes are saved and reflected whenever I re-open the edit view. Works as expected!

However, one thing I noticed was the following:

- Sign out > Sign in as instructor > Go to Courses (not course management) and select the course there > Select the exercise > Instructor access: Edit

Here, the competencies do not show up. Any idea why that is? They appear after a refresh of the page.

**Review by Gusti2010 — DISMISSED** [review]
Manually tested on ts11_Legacy. Works as expected!

**Review by pzdr7 — COMMENTED** [review]
Tested on TS2:

- Created and updated exercises (file upload, programming, modeling, text, quiz) in a course without competencies. Still works as expected.
- Adding / removing competencies to / from exercises:
  - Works as expected for file upload, programming, modeling, and text exercises
  - For quiz exercises, (de)selecting new competencies does not seem to mark the quiz as unsaved (without making further changes, I cannot save the quiz).

**Review by pzdr7 — APPROVED** [review]
Tested on TS2, works as expected for quizzes as well now!

**Review by georgyia — APPROVED** [review]
manually tested on ts2, works as expected

File: src/main/webapp/app/shared/competency-selection/competency-selection.component.ts [hunk 3]
```diff
@@ -21,14 +22,16 @@ import { finalize } from 'rxjs';
 export class CompetencySelectionComponent implements OnInit, ControlValueAccessor {
     @Input() labelName: string;
     @Input() labelTooltip: string;
-    @Input() value: any;
+    // selected competencies
+    @Input() value?: Competency[];
```
Comment (rstief) [inline src/main/webapp/app/shared/competency-selection/competency-selection.component.ts:26]: could rename to selectedCompetencies, also could avoid the undefined checks by setting its default value to `[ ]`

**Review by rstief — APPROVED** [review]
Code looks good, just a minor note for a possible change.

**Review by reschandreas — APPROVED** [review]
code lgtm

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**